Objects/unicodeobject.c

```diff
@@ -2272,34 +2269,79 @@ PyUnicode_FromString(const char *u)
     return PyUnicode_DecodeUTF8Stateful(u, (Py_ssize_t)size, NULL, NULL);
 }
 
+
 PyObject *
 _PyUnicode_FromId(_Py_Identifier *id)
 {
-    if (!id->object) {
-        id->object = PyUnicode_DecodeUTF8Stateful(id->string,
-                                                  strlen(id->string),
-                                                  NULL, NULL);
-        if (!id->object)
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+    struct _Py_unicode_ids *ids = &interp->unicode.ids;
+
+    Py_ssize_t index = id->index;
+    if (index < 0) {
+        struct _Py_unicode_runtime_ids *rt_ids = &interp->runtime->unicode_ids;
+
+        PyThread_acquire_lock(rt_ids->lock, WAIT_LOCK);
+        // Check again to detect concurrent access. Another thread can have
+        // initialized the index while this thread waited for the lock.
+        if (id->index < 0) {
```

Comment:
I'm not sure if the C compiler must always read again the new structure member. Using volatile avoids the need to ask such question, but I would prefer to avoid volatile if needed.

cc @pablogsal

Reply:
I added volatile. Using atomic variables is not possible right now: pycore_atomic.h remains an internal header, and I don't think that <stdatomic.h> is widely adopted yet.